log-levels.md

```diff
@@ -0,0 +1,46 @@
+# Log Levels 
+
+This guide serves as guideline for library authors with regards to what [SwiftLog](https://github.com/apple/swift-log) log levels are apropriate for use in libraries, and in what siuations to use what level.
+
+## Guidelines for Libraries
+
+SwiftLog defines the following 7 log levels: `trace`, `debug`, `info`, `notice`, `warning`, `error`, `critical`.
+
+Out of those, generally only levels _lower than_ info (exclusive) are generally okey to be used by libraries.
+In the followig section we'll explore how to use them in practice.
+
+### Recommended log levels
+
+It is always fine for a library to log at `trace` and `debug` levels, and these two should be the primary levels any library is logging at.
+
+As `trace` is the finest log level, and generally is not going to be used by end-users of a library unless debugging a very specific issue, it should be a way for developers of the library to "log everything we could possibly need to diagnose a hard to reproduce bug."
+
+**// TODO: should we suggest that a system should be able to survive with debug logging turned on in production (ktoso: I'd suggest that, yes).**
+
+This is in contrast with `debug` which is some users _may_ choose to run with enabled on their production systems. As such, debug logging should not be overly noisy, and provide a high value understanding of what is going on in the library for end users, without diving deep into its internals (which `trace` is intended for).
+
+### Log levels to avoid
+
+All these rules are only _general_ guidelines, and as such may have exceptions to the rules. Please consider the following examples and rationale for why logging at high log levels by a library may not be desirable:
+
+It is generally _not acceptable_ for a service client (e.g. an http client) to log an `error` when a request has failed. End-users may be using the client to probe if an endpoint is even responsive or not, and a failure to respond may be _expected_ behavior for the end-user, and logging errors would only confuse and pollute their logs.
+Instead, libraries should either `throw`, or return an `Error` value that users of the library will have enough knowladge about if they should log or ignore it.
+
+It is even less acceptable for a library to log any successful operations. This leads to flodding server side systems, especially 
+
+### Suggested logging style
```

Comment:
one other thing worth explaining is performance / cost of computing logging metadata. some library authors may not pay enough attention and / or release that they can take advantage of the auto-closure nature of swift-log so we should highlight that